ENH Model Card

merveenoyan opened the pull request.

adrinjalali commented on skops/saving_utils/default_template.md:
```diff
@@ -0,0 +1,27 @@
+
+## Intended uses & limitations
+
+This model doesn't exist, so you probably don't want to use it! This is just an example template. Please write a very thoughtful model card ❤️
```
make sure your lines are not too long. If you're using VSCode, the Rewrap extension is your friend :)

adrinjalali commented on skops/saving_utils/model_card.py:
```diff
@@ -0,0 +1,55 @@
+
+
+def _plot_estimator(model):
+    set_config(display="diagram")
```
We should avoid setting any config on the global `sklearn` level, since it would interact and conflict with what the user has set. You can use `config_context` instead: https://scikit-learn.org/stable/modules/generated/sklearn.config_context.html#sklearn.config_context

merveenoyan replied:
yeah I thought about setting config to different values at beginning and end of function itself which was dirty, thanks for pointing out!

adrinjalali commented on skops/saving_utils/model_card.py:
```diff
@@ -0,0 +1,55 @@
+
+def _plot_estimator(model):
+    set_config(display="diagram")
+    return estimator_html_repr(model)
```
Do you need to set the config when you're explicitly calling this function anyway?

merveenoyan replied:
OK apparently that's what it does anyway, I'll directly return that :)

merveenoyan replied:
do you think it's good to give the people the option to have it ask JSON? @adrinjalali

adrinjalali replied:
yeah that certainly wouldn't hurt.

adrinjalali commented on skops/saving_utils/model_card.py:
```diff
@@ -0,0 +1,55 @@
+def _extract_estimator_config(model):
+    """Extracts estimator configuration and renders them into a vertical table.
+
+    Args:
+        model (estimator): scikit-learn pipeline or model.
+
+    Returns:
+        str: Markdown table of hyperparameters.
```
This repo uses `rst` format with numpydoc style. This section would be:

```
    Parameters
    ----------
    model: estimator
        scikit-learn compatible estimator.

    Returns
    -------
    str:
        Markdown table of hyperparameters.
```

merveenoyan replied:
I did pre-commit install which I thought would automatically format things but apparently it's not the case 😂

adrinjalali commented on skops/saving_utils/model_card.py:
```diff
@@ -0,0 +1,55 @@
+    return estimator_html_repr(model)
+
+
+def _extract_estimator_config(model):
```
this should be tested for cases when the huperparameters are non-trivial, like a custom scorer passed to a GridSearchCV

merveenoyan replied:
just to confirm what I understood, Grid Search CV object itself returns best estimator no? How is it different? what do you mean with custom scorer? 😅

adrinjalali replied:
`GridSeachCV` doesn't return the best estimator. It stores the best estimator in the `best_estimator_` attribute and uses it when you call `predict` on a `GridSearchCV` instance. For instance, you can see how the parameters of the `gs` here look like: https://scikit-learn.org/stable/auto_examples/model_selection/plot_multi_metric_evaluation.html#running-gridsearchcv-using-multiple-evaluation-metrics

merveenoyan replied:
Ah maybe I remembered it like that because that's how I always use it 😅

merveenoyan replied:
it still returns stuff we can parse, no? see below:

```
gs.get_params()
{'cv': None,
 'error_score': nan,
 'estimator': DecisionTreeClassifier(random_state=42),
 'estimator__ccp_alpha': 0.0,
 'estimator__class_weight': None,
 'estimator__criterion': 'gini',
 'estimator__max_depth': None,
 'estimator__max_features': None,
 'estimator__max_leaf_nodes': None,
 'estimator__min_impurity_decrease': 0.0,
 'estimator__min_samples_leaf': 1,
 'estimator__min_samples_split': 2,
 'estimator__min_weight_fraction_leaf': 0.0,
 'estimator__random_state': 42,
 'estimator__splitter': 'best',
 'n_jobs': 2,
 'param_grid': {'min_samples_split': range(2, 403, 20)},
 'pre_dispatch': '2*n_jobs',
 'refit': 'AUC',
 'return_train_score': True,
 'scoring': {'AUC': 'roc_auc', 'Accuracy': make_scorer(accuracy_score)},
 'verbose': 0}
```

adrinjalali replied:
we can parse and persist them, but the `make_scorer(accuracy_score)` is code, it's not a literal or a number or the usual stuff you'd expect. We just need to make sure such a case is tested with the way you're creating the table.

adrinjalali reviewed (commented):
Thanks for starting this @merveenoyan 

I'd put the file under `skops/card/_model_card.py` and expose the main function through `skops/card/__init__.py`

merveenoyan commented:
lol I'll never do `git add .` again 😂
it was like I always check `git status` and add stuff accordingly, this time pre-commit annoyed me so I did `git add .` and look where it got me

adrinjalali commented:
> lol I'll never do `git add .` again joy it was like I always check `git status` and add stuff accordingly, this time pre-commit annoyed me so I did `git add .` and look where it got me

What was the issue?

adrinjalali commented:
I'd say the user should create the modelcard separately, and pass the object, or file name to `init`. There should be quite a bit of content in the modelcard which is written by the user, it doesn't make sense for `init` to call the modelcard creation then.

merveenoyan commented:
Model card is already created in target directory where repository is (so you so init(), create the card and push) (if the user follows the example or doesn't get confused 😅)

merveenoyan commented:
@adrinjalali I haven't added tests yet, you can roughly check it out?

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -0,0 +1,60 @@
+    return table
+
+
+def create_model_card(
+    model, path, model_id=None, license=None, tags=None, metrics=None
+):
```
I'm not sure about the API.

How is the user supposed to fill out sections which are text? I would expect an interface which would allow the user to insert all the sections that we're talking about with others in the modelcard team.

So maybe the API would be more like:

```python
card = skops.card.model_card_factory(...)
card.limitations = ...
card.author = ...
...
card.save(path/to/README.md, overwrite=True)
```

WDYT?

also pinging @nateraw here.

merveenoyan replied:
isn't factory a bit overkill 😅 as for attributes, I agree.

merveenoyan replied:
If we'll go with this I have questions:
1. Are we going with factory? If so, why? Can't we simply return default jinja template and let user inject values? 
2. How does output of model card factory differ from `default_template.md` (as a jinja template and not as a string)?

adrinjalali replied:
To answer these questions, I think it makes sense for you to write down how you think the end user would interact with the object and functions we have in an example.

Whatever we choose, should be based on how the end user interacts with the library.

So maybe as the next step write a `plot_model_card.py` example with everything you thin the user can do, and we take it from there?

merveenoyan replied:
I thought of the one I've already written in the `plot_hf_hub.py` but I'll write a more in depth one :)

nateraw replied:
can't this `create_model_card` fn just take in the template kwargs if you want to override them? like limitations, authors, etc? Right now its not possible to override these after the class has been instantiated within `from_template`.  Also unsure if this card creation will be user-facing if its just going to be created automatically under the hood anyways.

adrinjalali reviewed (commented):
You'd need to add modelcards to the dependencies for the CI to pass and the docs to build.

nateraw commented on skops/card/_model_card.py:
```diff
@@ -0,0 +1,60 @@
+        model_plot=model_plot,
+    )
+
+    with open(f"{path}/README.md", "w+", encoding="utf-8") as f:
```
Why not `card.save(f"{path}/README.md")`?

nateraw reviewed (commented):
Few comments as you're working through this. Looking pretty good so far :) Thanks for digging into `modelcards` @merveenoyan

adrinjalali commented on skops/card/default_template.md:
```diff
@@ -0,0 +1,32 @@
+The model plot is below.
+
+{% autoescape false %}
+  {{ model_plot | replace("\n             ", "")}}
```
I'm not sure if it's a stable solution though, the spaces might simply change.

merveenoyan replied:
replaced with regex

merveenoyan commented:
@adrinjalali does something like below work? 

```
model_card = card.create_model_card(gs, local_repo)
model_card.author = "merve"
model_card.limitations = "This model is a dummy one"
model_card.save(f"{local_repo}/README.md")
```

that is, I return the card instead of saving it:
```
def create_model_card(
    model, path, model_id=None, license=None, tags=None, metrics=None
):
    ...
    card = ModelCard.from_template(
        card_data=CardData(
            license=license,
            library_name="sklearn",
            tags=tags,
            metrics=metrics,
        ),
        template_path="skops/skops/card/default_template.md",
        model_id=model_id,
        hyperparameter_table=hyperparameter_table,
        model_plot=model_plot,
    )
    
    return card
```
I think it's simple & still does what you want it to do.

merveenoyan commented on examples/plot_model_card.py:
```diff
@@ -0,0 +1,57 @@
+
+model_card = card.create_model_card(model)
+model_card.limitations = "This model is not ready to be used in production."
+model_card.license = "afl3.0"
```
@nateraw I couldn't figure out a way to set it like this as the way the library currently is, what am I doing wrong? might be missing a very stupid thing😅

nateraw replied:
The license would be within the card data, which is at card.data

merveenoyan commented:
I think I got it right this time @nateraw now writing the tests @adrinjalali

merveenoyan commented:
apparently `regex_replace` is not one of builtin jinja filters, I'll find another way.

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -0,0 +1,59 @@
+scikit-learn compatible model and save it.
+"""
+
+# %% Imports
+# =======
```
```suggestion
# %%
# Imports
# =======
```

same for the other sections

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -0,0 +1,59 @@
+
+# %% Data
+# ====
+# Then we create some random data to train and evaluate our model.
```
this is not random data, you're loading the breast cancer dataset.

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -0,0 +1,66 @@
+    model_plot = str(estimator_html_repr(model))
+    hyperparameter_table = _extract_estimator_config(model)
+    if card_data is None:
+        card_data = CardData(library_name="sklearn")
```
should we set the `library_name` regardless of what's provided in `card_data`?

merveenoyan replied:
yes I will add it, good catch

adrinjalali reviewed (commented):
Not a complete review, the PR is getting there.

Are you going to add the other visuals as well (model interpretation etc)?

Those could also be separate PRs.

adrinjalali commented:
You need to merge with the latest `main` and fix the dependencies as well.

merveenoyan commented on skops/card/default_template.md:
```diff
@@ -0,0 +1,27 @@
+The model plot is below.
+
+{% autoescape false %}
+  {{ model_plot | regex_replace("\n\s+", "")}}
```
Makes sense, IDK why I couldn't see it 🤨

merveenoyan commented:
Resulting model card as of current PR: https://huggingface.co/merve/sklearn-9e3be5a1-538f-4db6-b5e3-27d02a310b12 @adrinjalali addressed your comments, sorry I was a bit careless with this PR, it won't happen again.

merveenoyan commented:
@nateraw I have an error for windows env where the model card saving raises an encoding error, I looked at the card module and it doesn't give me option to set encoding. (every test except for windows ones pass) 
@adrinjalali do you know a better way of setting path to template? readthedocs build can't find it.

BenjaminBossan commented:
> I have an error for windows env where the model card saving raises an encoding error

Not sure if that's the cause, but I saw that you use `/` to denote folders, but in Windows, it's `\`. The most robust way should be to use something like:

```python
import os
template_path = os.path.join("skops", "card", "default_template.md")
```

merveenoyan commented:
@BenjaminBossan I google'd it and it was about encoding so I didn't fix it like that, now that I fix it I still get the same error, maybe it's about encoding 😅

BenjaminBossan commented:
I see. Since I know basically nothing about this project right now, I can only make guesses :)

Regarding encoding, maybe it's this line?

> C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\modelcards\cards.py:82: in save
>    filepath.write_text(str(self))

The [`write_text`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.write_text) method of pathlib allows to indicate the encoding, so maybe that needs to be done. Of course, this would require a change in modelcards.

adrinjalali commented:
I tend to set the path with pathlib.Path as `Path(blah) / blah / blah`, much easier for me to read.

But regarding encoding, if we have a minimal reproducible, we can fix the modelcard part. @nateraw who mostly has been writing that code also works at HF. He's planning to merge that code with the `huggingface_hub` this week, so we can fix it before that's done.

nateraw commented:
To be clear, we need to add encoding to `modelcards` (as you originally did before using `card.save` @merveenoyan )? I can do that today and issue a release if that helps.

nateraw commented on skops/_min_dependencies.py:
```diff
@@ -12,6 +12,7 @@
 dependent_packages = {
     "scikit-learn": ("0.24", "install", None),
     "huggingface_hub": ("0.8.1", "install", None),
+    "modelcards": ("0.1.4", "install", None),
```
```suggestion
    "modelcards": ("0.1.6", "install", None),
```

nateraw commented on skops/card/tests/test_card.py:
```diff
@@ -0,0 +1,69 @@
+        model_card = create_model_card(
+            model, card_data=card_data, model_description="sklearn FTW"
+        )
+        model_card.save(os.path.join(f"{dir_path}", "README.md"))
```
why not pathlib 😢 - pathlib brings joy to many.

nateraw reviewed (commented):
I released `modelcards==0.1.6` just now. Lets see how that goes.

merveenoyan commented:
@adrinjalali do you know where readthedocs builder is running from? it can't find the default template for some reason.

BenjaminBossan commented:
> do you know where readthedocs builder is running from? it can't find the default template for some reason.

Not Adrin, but may still have an answer :)

It could be that readthedocs searches in the wrong location, but there is also the possibility that the `.md` file is [not bundled with the Python package](https://python-packaging.readthedocs.io/en/latest/non-code-files.html). In the latter case, it might need to be included in the MANIFEST.in (see [these docs](https://packaging.python.org/en/latest/guides/using-manifest-in/)) and/or added in the `setup` call inside `setup.py` (see [these docs](http://peak.telecommunity.com/DevCenter/setuptools#including-data-files)).

nateraw commented:
@BenjaminBossan is absolutely right! Thank you ❤️

Beware - the tests can deceive you here. Without including the card in the package, it won't be found for folks who `pip install`'d this.

@merveenoyan tl;dr just follow what I did in `modelcards`:

- Add a [MANIFEST.in](https://github.com/nateraw/modelcards/blob/main/MANIFEST.in) in the root of the repo
- `include_package_data=True` in [setup.py](https://github.com/nateraw/modelcards/blob/7cb1c427a75c0a796631c137c5690ceedab3b2f8/setup.py#L28)

I'm actually not 100% sure you need to do both of these things, but I'm just going to trust that when I did it, it was for a reason lol 😅  . I could be wrong, but IIRC one of these works with a certain type of build (ex. egg, bdist, sdist) and the other works for another...without both it can break depending on how you install/distribute the package.

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -0,0 +1,58 @@
+        scikit-learn pipeline or model.
+    card_data: CardData
+        CardData object. See the
+        [docs](https://github.com/nateraw/modelcards/blob/main/modelcards/card_data.py#L78).
```
IMHO, it's a bit dangerous to add a link to a specific line of code here. This information will become outdated easily when the file it points to is changed. Ideally, it should link to the documentation, but AFAICT, there is none.

I see two alternatives here:

1. Create a permalink. This can be achieved by visiting the github page and then pressing `y`. The URL bar will now show the [permalink](https://github.com/nateraw/modelcards/blob/7cb1c427a75c0a796631c137c5690ceedab3b2f8/modelcards/card_data.py#L78). This way, this line will always point to this specific version of the file, so there is no danger of it suddenly pointing to something else (or nothing). The disadvantage is that if the class is updated, this link does not reflect it.
2. It's a bit hacky, but you could also add a link to the search for "class CardData" like so: https://github.com/nateraw/modelcards/search?q=%22class+CardData%22. The advantage is that if the class is moved, the search will still find it. But if the class is renamed or if there are more classes with the same name, it won't work anymore.

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -0,0 +1,83 @@
+model_card = card.create_model_card(
+    model,
+    card_data=card_data,
+    template_path="../skops/card/default_template.md",
```
I don't think users of `skop` should necessarily choose the template. By default we should use the template provided in the package.

adrinjalali replied:
this is still here.

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -0,0 +1,83 @@
+)
+save_dir = mkdtemp(prefix="skops")
+
+model_card.save(f"{save_dir}/README.md")
```
we should explain here that they should use `init` to initialize a folder for a model, and dump this modelcard in the same folder. Maybe we could do the same in this example just to showcase?

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -0,0 +1,83 @@
+# We first set the metadata with CardData and pass it to create_model_card.
+# We pass information other than metadata in kwargs.
+
+limitations = "This model is not ready to be used in production."
+model_description = (
+    "This is a HistGradientBoostingClassifier model trained on breast cancer dataset."
+    " It's trained with Halving Grid Search Cross Validation, with parameter grids on"
+    " max_leaf_nodes and max_depth."
+)
+license = "mit"
+card_data = CardData(
+    license=license,
+    tags=["tabular-classification"],
+    datasets="breast-cancer",
+    metrics=["acc"],
+)
```
I think we need to also add the other sections that we've agreed on for the default template, like author, citations, etc.

adrinjalali commented on skops/card/tests/test_card.py:
```diff
@@ -0,0 +1,69 @@
+
+
+def fit_model():
+    import numpy as np
+    from sklearn.linear_model import LinearRegression
```
you can move the imports to the top of the file.

adrinjalali commented on skops/card/tests/test_card.py:
```diff
@@ -0,0 +1,69 @@
+    return reg
+
+
+def write_card():
```
maybe `generate_card` instead? or `get_card`?

adrinjalali commented on skops/card/tests/test_card.py:
```diff
@@ -0,0 +1,69 @@
+from skops.card import create_model_card
+
+
+def _get_cwd():
```
you don't seem to be using this function here, if not, you could remove it.

adrinjalali reviewed (commented):
This is looking pretty good. A few small points, and other than the packaging issue, looks great.

Let me know if you need help with the packaging of the template file.

merveenoyan commented:
@adrinjalali I addressed your comments.

merveenoyan commented:
@adrinjalali I added citation info, how to use the model code, author, model card contact sections. The resulting model card looks like [this](https://huggingface.co/merve/sklearn-b893d4d5-2e7c-4f0a-8295-46e88d8aaad1).

adrinjalali commented:
This doesn't look right:

![image](https://user-images.githubusercontent.com/1663864/178302851-30963581-84fd-4103-b911-2c6b1cfe6be8.png)


Also, should we put the hyperparameters in a `<details>` tag as well? They are quite long.

adrinjalali commented:
As for optional sections, should we remove "Optional" from the title text, and only have them if they actually exist? Or at least remove "Optional" maybe? At the end of the day, everything is kind of optional.

BenjaminBossan commented:
@merveenoyan Sorry I pasted a mistake in my code:

```python
- ROOT = skops.__path__
+ ROOT = skops.__path__[0]
```

Btw. to test it without waiting for CI, you could install the package in a local environment and then check inside of site-packages if all necessary files can be found.

merveenoyan commented:
@BenjaminBossan I noticed it before I saw your comment and did something like below:
```
if isinstance(ROOT, list):
        ROOT = ROOT[0]
```
but [apparently it always returns list](https://stackoverflow.com/questions/22117335/why-is-path-a-list) so I'll change it 😅

merveenoyan commented:
When I checked my local black version, it was 21.12b0 then I looked at pre-commit config and saw it was 22.3.0 so I updated it, but IDK why now everything's up to date, and I get style error in tests 🤨 I'm checking

merveenoyan commented:
@adrinjalali put hyperparameters inside the expandable section and put code inside backticks to format [here](https://huggingface.co/merve/sklearn-d6727038-2a68-4a20-a675-b4d36e8e4855)

merveenoyan commented:
@adrinjalali there was something going on with CI yesterday so I triggered it again and all the tests pass, can you check?

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -0,0 +1,106 @@
+
+from skops import card, hub_utils
+
+# %% Data
+# ====
```
```suggestion
# %% 
# Data
# ====
```

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -0,0 +1,106 @@
+print("X's summary: ", X.describe())
+print("y's summary: ", y.describe())
+
+# %% Train a Model
+# =============
```
```suggestion
# %%
# Train a Model
# =============
```

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -0,0 +1,106 @@
+).fit(X_train, y_train)
+model.score(X_test, y_test)
+
+# %% Create a model card
+# =======================
```
```suggestion
# %%
# Create a model card
# ====================
```

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -0,0 +1,62 @@
+    Parameters:
+    ----------
+    model: estimator
+        scikit-learn pipeline or model.
```
```suggestion
        scikit-learn compatible estimator.
```

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -0,0 +1,62 @@
+        scikit-learn pipeline or model.
+    card_data: CardData
+        CardData object. See the
+        [docs](https://github.com/nateraw/modelcards/blob/7cb1c427a75c0a796631c137c5690ceedab3b2f8/modelcards/card_data.py#L78).
```
we shouldn't refer to a github file. This should reference a documentation page.

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -0,0 +1,62 @@
+        CardData object. See the
+        [docs](https://github.com/nateraw/modelcards/blob/7cb1c427a75c0a796631c137c5690ceedab3b2f8/modelcards/card_data.py#L78).
+    card_kwargs:
+        Card kwargs are information you can pass to fill in the sections of the
+        card template, e.g. description of model
```
what are the available ones?

adrinjalali commented:
Does the example we have cover all available fields?

We need somewhere in the docs to explain what all available fields are, and what they mean and how they should be filled. We don't seem to have it anywhere ATM.

adrinjalali commented:
Would you like us to merge this one, and then have another PR where you work on the documentation? I think the API makes sense as is.
